snap/get_version

```diff
@@ -0,0 +1,116 @@
+#!/bin/sh
+set -eu
+
+# Script to build the version to set to the snap.
+
+# When considering a tag, if starting with "<current branch name>-" it will have its prefix removed.
+# For instance:
+# * 0.1 -> 0.1
+# * msentraid-0.1 -> 0.1 on msentraid branch.
+
+# 1. If current commit is tagged, the version is directly the tag name.
+# 2. If current commit is not tagged, the version is:
+#    a. <last tag name on current branch>+<commit_sha> for main branch.
+#    b. <last tag name on current branch (not in main)>+<commit_sha>.<last_commit_merged_from_main> for other branches.
+#
+# Any of those version will be annoted with +dirty if there are local changes.
+
+# set_version will markup the version in the snapcraft.yaml file after amending it with a dirty markup if necessary.
+# $1: version: the version to set.
+set_version() {
+    version="${1}"
+
+    version=$(annotate_with_dirty "${version}")
+    craftctl set version="${version}"
+}
+
+# annotate_with_dirty may amend the version with a dirty markup if there are local changes.
+# $1: version: the version to annotate.
+annotate_with_dirty() {
+    version="${1}"
+
+    # check if current tree content is dirty.
+    is_dirty=$(git -C "${SNAPCRAFT_PART_SRC}" status --porcelain)
+    if [ -n "${is_dirty}" ]; then
+        version="${version}+dirty"
+    fi
+
+    echo "${version}"
+}
+
+# last_tag_on_branch returns the last tag on that branch that is not in another provided branch.
+# $1: origin_branch: branch we start listing from.
+# $2: other_branch: the branch where we don’t want to tags to be reachable from.
+# If we don’t have another branch to compare with, it will just return the most recent tag.
+last_tag_on_branch() {
+    origin_branch="${1}"
+    other_branch="${2}"
+
+    all_tags_on_origin_branch=$(git -C "${SNAPCRAFT_PART_SRC}" tag --sort=-committerdate --merged "${origin_branch}")
```

Comment:
This lists the oldest tag first, so this function returns the oldest commit, not the latest

Reply:
Are you sure about this? `--sort=committerdate` would sort with the older commits first, but `--sort=-committerdate` defintively lists the more recent ones first. This is also what I can verify with my manual tests.

Reply:
I also tested it and in my tests it definitely doesn't list the most recent commit first. Sorting by version works though:

```
git tag --sort=-v:refname
```

Reply:
The issue with refname is that we shouldn’t really on the version ascii reference.

But I just saw that you resolve it, so this is working now for you as intended?

Reply:
No, it doesn't work for me. Not sure why the conversation was resolved, must have been by accident.

Reply:
> The issue with refname is that we shouldn’t really on the version ascii reference.

Why not? We don't plan to ever go down with the version number, right? And `v:refname` sorts versions correctly, it doesn't just do compare the ASCII characters. So for example it correctly sorts `0.10.0` as a higher version than `0.9.0`.

Reply:
Can you share the output of your sort command on the current repo, which has 0.1, msentraid-0.1 (and you can add a newer commit on main and msentraid branch to test).

To me, this is giving:
```
$ git -C . tag --sort=-committerdate --merged msentraid
msentraid-0.2test
v0.2test
msentraid-0.1
0.1
```
Which is the order I expected by recency.

I’m fine going with the refname if you think they will always make logical sense. However, I want to ensure we are not missing anything as you clearly have different results on the sort parameter already.

Reply:
> Can you share the output of your sort command on the current repo, which has 0.1, msentraid-0.1 (and you can add a newer commit on main and msentraid branch to test).

Sure, lets look at a snippet of the output in the `git.launchpad.net/ubuntu/+source/apport` repo (on branch `ubuntu/noble`):

```
$ git tag --sort=-committerdate
applied/0.1
applied/0.10
applied/0.100
applied/0.101
<snip>
applied/0.18
applied/0.19
applied/0.2
applied/0.20
applied/0.21
<snip>
```

The `applied/0.1` is not the latest tag, and it was not created after the `applied/0.10` tag.

Reply:
Indeed, I don’t have tangible explanation looking at committer and author dates on those. I don’t know how you ended up trying this on apport git-ubuntu history to find such an example, but that’s ok.

Reply:
FTR, I tried it on various git repos, apport was just an example, and it lists the tags in an unexpected order in almost all of them.